File: python/day17.py
from aoc import *


left = -1
right = 1
down = -1j

rocks = [
    {0, 1, 2, 3},
    {1, 1j, 1+1j, 2+1j, 1+2j},
    {0, 1, 2, 2+1j, 2+2j},
    {0, 1j, 2j, 3j},
    {0, 1, 1j, 1+1j}
]


def move(rock, delta):
    return set(map(lambda pt: pt + delta, rock))

def is_inbounds(rock):
    return all(0 <= pt.real < 7 for pt in rock)

def would_clash(rock, tower):
    return any(pt in tower for pt in rock)

def peaks(tower, max_y):
    return frozenset(pt.real for pt in tower if pt.imag == max_y)


def play(movements, rounds):
    M = len(movements)
    R = len(rocks)
    m = 0
    max_y = 0
    extra_y = 0
    seen = dict()
    tower = set(range(7))
    t = 0
    while t < rounds:
        rock = move(rocks[t%R], 2 + (4+max_y)*1j)
        t += 1
        while True:
            dx = [left, right][movements[m%M] == '>']
            m += 1
            rock_ = move(rock, dx)
            if is_inbounds(rock_) and not would_clash(rock_, tower):
                rock = rock_
            rock_ = move(rock, down)
            if not would_clash(rock_, tower):
                rock = rock_
            else:
                max_y = int(max(max_y, *(pt.imag for pt in rock)))
                tower = rock | {pt for pt in tower if pt.imag >= max_y - 100}
                t_hash = peaks(tower, max_y), t%R, m%M
                if t_hash in seen:
                    t1, y1 = seen[t_hash]
                    dt = t - t1
                    dy = max_y - y1
                    skipped_periods = (rounds - t) // dt
                    extra_y += skipped_periods * dy
                    t += skipped_periods * dt
                seen[t_hash] = t, max_y
                break
    return max_y + extra_y


def solve(filename=17):
    movements = read_input_line(filename)
    return play(movements, 2022), play(movements, 1000000000000)

print(solve())










def test():
    assert solve("17_test") == (3068, 1514285714288)
    assert solve() == (3232, 1585632183915)

File: clojure/day17.clj
(ns day17
  (:require aoc))


(defrecord Point [x y])


(defmacro ->Rock [pts]
  `(map (fn [[x# y#]] (->Point x# y#)) ~pts))

(def rocks
  [(->Rock [[0 0] [1 0] [2 0] [3 0]])
   (->Rock [[1 0] [0 1] [1 1] [2 1] [1 2]])
   (->Rock [[0 0] [1 0] [2 0] [2 1] [2 2]])
   (->Rock [[0 0] [0 1] [0 2] [0 3]])
   (->Rock [[0 0] [1 0] [0 1] [1 1]])])

(defn move-left  [{:keys [x y]}] (->Point (dec x) y))
(defn move-right [{:keys [x y]}] (->Point (inc x) y))
(defn move-down  [{:keys [x y]}] (->Point x (dec y)))

(defn pt+ [a b]
  (->Point (+ (:x a) (:x b)) (+ (:y a) (:y b))))

(defn inbounds? [rock]
  (reduce
   (fn [acc x]
     (if-not (<= 0 x 6) (reduced false) acc))
   true
   (map :x rock)))

(defn peaks-hash [tower max-y]
  (transduce
   (map (fn [{:keys [x y]}]
      (if (= y max-y) (* x x) 0)))
   +
   tower))


(defn play [movements rounds]
  (let [M (count movements)
        R (count rocks)
        tower (set (for [x (range 7)] (->Point x 0)))]
    (loop [[r m max-y skipped seen tower]
           [0 0 0 0 (transient {}) tower]]
      (if (>= r rounds) (+ max-y skipped)
          (let [init-pos (->Point 2 (+ max-y 4))
                rock     (mapv (partial pt+ init-pos) (nth rocks (mod r R)))]
            (recur
             (loop [rock rock
                    m    (long m)]
               (let [move-fn    (if (= \< (nth movements (mod m M)))
                                  move-left move-right)
                     rock'      (mapv move-fn rock)
                     moved-rock (if (and (inbounds? rock')
                                         (not-any? tower rock'))
                                  rock' rock)
                     rock'' (mapv move-down moved-rock)]
                 (if (not-any? tower rock'')
                   (recur rock'' (inc m))
                   (let [max-y       (max max-y (:y (peek moved-rock)))
                         tower       (into (set moved-rock)
                                           (filter #(> (:y %) (- max-y 100)))
                                           tower)
                         t-hash      (+ ^long (peaks-hash tower max-y)
                                        ^long (* 100 (mod r R))
                                        ^long (* 500 (mod m M)))
                         r           (inc r)
                         [r skipped] (if-let [[r1 y1] (seen t-hash)]
                                       (let [dr (- r r1)
                                             dy (- max-y y1)
                                             skipped-periods (quot (- rounds r) dr)]
                                         [(+ r (* skipped-periods dr))
                                          (+ skipped (* skipped-periods dy))])
                                       [r skipped])]
                     [r (inc m) max-y skipped (assoc! seen t-hash [r max-y]) tower]))))))))))


(defn solve
  ([] (solve 17))
  ([input]
   (let [movements (aoc/read-input-line input)]
     [(play movements 2022)
      (play movements 1000000000000)])))


(solve)
